### PR TITLE
Voter dApp to store committed salts in cookies

### LINKED ISSUE #1476: Persist commit salt in voting clients
Plus also remind user to back up salt. If user has salt then they can reveal votes manually in case a client has a bug

### PR DESCRIPTION
**Pull request by nicholaspai**
Resolves #1476

Stores and displays all prices & salts before they are encrypted for the user. Whether user submits transaction or not, the price & salt are stored locally to a cookie. Adds a tool-tip explaining why this data should be kept as a backup in case the dApp fails to encrypt-decrypt properly. Example display and logging of browser cookies is shown. 

One side effect of this change is that all ethereum accounts who use the same browser have access to each other's saved salts. I anticipate this not being a problem for the majority of users who do not share devices with others to sign transactions, but it is something to be aware of.

Future work would be to track the `batchCommitFunctions` and only store commit-data of votes to cookies that the user has signed and submitted (`commitStatus === pending/success`).

Screen Shot 2020-05-28 at 17 54 29

Uses this [react cookie library](https://github.com/reactivestack/cookies/tree/master/packages/react-cookie)

I also added a helper script, `RevealVote.js` that I used to successfully reveal a vote via a script using the salt displayed on the dApp.
 
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

### CODE REVIEW

**Comment by chrismaree** [comment]
This looks awesome! Is the last commit salt column always shown, as in your screenshot? Also, that popover is really great; we should consider adding that in a few other complicated parts of the dapp.

File: core/scripts/local/RevealVote.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/**
```
Comment (chrismaree) [inline core/scripts/local/RevealVote.js:1]: +1 another useful script

Side note: we've not got a bunch of these scripts you've created for doing atomic action on our protocol. Should we re-consider the directory structure for these? the `local` directory is getting a bit crowded and has a bunch of unrelated things in it.

File: voter-dapp/src/ActiveRequests.js [hunk 2]
```diff
@@ -423,6 +487,7 @@ function ActiveRequests({ votingAccount, votingGateway }) {
                 <TableCell>{formatDate(pendingRequest.time, drizzle.web3)}</TableCell>
                 <TableCell>{statusDetails[index].statusString}</TableCell>
                 <TableCell>{getCurrentVoteCell(index)}</TableCell>
+                <TableCell>{getLastCommittedSalt(index)}</TableCell>
```
Comment (chrismaree) [inline voter-dapp/src/ActiveRequests.js:490]: how does this render on smaller screens? did you test it out on mobile?

File: core/scripts/local/RevealVote.js (hunk 1 above)
Comment (nicholaspai) [reply core/scripts/local/RevealVote.js:1]: Yes we should group them into `/emp` and `/dvm` for example but that's outside the scope of this PR imo. I'll make an issue

**Comment by nicholaspai** [comment]
> This looks awesome! Is the last commit salt column always shown, as in your screenshot? Also, that popover is really great; we should consider adding that in a few other complicated parts of the dapp.

Last commit salt is only shown once the user has a cookie, otherwise its blank

File: voter-dapp/src/ActiveRequests.js (hunk 2 above)
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js]: It does not render well, but the voter dApp in general does not render well on mobile. I think displaying the full salt is overkill so I decided to truncate the salt to the first 6 digits and added a "CopyToClipboard" button for the user's convenience 

Screen Shot 2020-05-28 at 09 53 16

**Comment by nicholaspai** [comment]
> First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> 
> What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> 
> The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> 
> One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> 
> Thoughts?

I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?

**Comment by mrice32** [comment]
> > First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> > What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> > The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> > One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> > Thoughts?
> 
> I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?

Isn't the idea of the backup that it protects you in case they do differ for some reason or the decryption fails? Or maybe I'm misunderstanding?

**Comment by nicholaspai** [comment]
> > > First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> > > What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> > > The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> > > One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> > > Thoughts?
> > 
> > 
> > I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?
> 
> Isn't the idea of the backup that it protects you in case they do differ for some reason or the decryption fails? Or maybe I'm misunderstanding?

I agree with you that the backup of the salt protects you in case the decryption (during the reveal phase) fails. This is what happened to me: the dApp did not decrypt the correct salt for my commit and therefore I couldn't reveal. What I would have liked to have done would be to use the salt to manually reveal through etherscan or a script.

**Comment by mrice32** [comment]
> > > > First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> > > > What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> > > > The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> > > > One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> > > > Thoughts?
> > > 
> > > 
> > > I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?
> > 
> > 
> > Isn't the idea of the backup that it protects you in case they do differ for some reason or the decryption fails? Or maybe I'm misunderstanding?
> 
> I agree with you that the backup of the salt protects you in case the decryption (during the reveal phase) fails. This is what happened to me: the dApp did not decrypt the correct salt for my commit and therefore I couldn't reveal. What I would have liked to have done would be to use the salt to manually reveal through etherscan or a script.

I totally agree that having a backup salt would've been helpful in your case. I think we should be storing them as you're currently doing, but I think the presentation to the user is currently a little confusing. I think it's confusing because the backups are mixed in with other information (like the decrypted price) that the dApp *is* actively using. This means that on the same line we're showing the *decrypted* price and the *backup* salt, leaving out the *decrypted* salt and the (nonexistent) *backup* price.

I was thinking an alternative might be a button on each line of the table that you click to show your local salt/price backups in a modal (we could even show all the backups you've ever made for that request. Then, we could make it obvious that if you want to use these, then you'll need to do the reveal manually.

Another, simpler option would be to just re-label the column as local salt backup, or something like that and explain in the tooltip that the dApp isn't using this salt.

**Comment by nicholaspai** [comment]
> > > > > First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> > > > > What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> > > > > The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> > > > > One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> > > > > Thoughts?
> > > > 
> > > > 
> > > > I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?
> > > 
> > > 
> > > Isn't the idea of the backup that it protects you in case they do differ for some reason or the decryption fails? Or maybe I'm misunderstanding?
> > 
> > 
> > I agree with you that the backup of the salt protects you in case the decryption (during the reveal phase) fails. This is what happened to me: the dApp did not decrypt the correct salt for my commit and therefore I couldn't reveal. What I would have liked to have done would be to use the salt to manually reveal through etherscan or a script.
> 
> I totally agree that having a backup salt would've been helpful in your case. I think we should be storing them as you're currently doing, but I think the presentation to the user is currently a little confusing. I think it's confusing because the backups are mixed in with other information (like the decrypted price) that the dApp _is_ actively using. This means that on the same line we're showing the _decrypted_ price and the _backup_ salt, leaving out the _decrypted_ salt and the (nonexistent) _backup_ price.
> 
> I was thinking an alternative might be a button on each line of the table that you click to show your local salt/price backups in a modal (we could even show all the backups you've ever made for that request. Then, we could make it obvious that if you want to use these, then you'll need to do the reveal manually.
> 
> Another, simpler option would be to just re-label the column as local salt backup, or something like that and explain in the tooltip that the dApp isn't using this salt.

I see what you're saying. I now save a local copy of the committed price and salt which should be available even if the app cannot decrypt the vote on-chain. I opted to still just save the latest committed data, which will only update if the `batchCommitFunction` succeeds.

Screen Shot 2020-05-28 at 15 24 17

@chrismaree unfortunately I had to scrap the copy button, as nice as it was UI-wise, because for some reason it doesn't work the way I intend it to

**Comment by mrice32** [comment]
> > > > > > First of all, this is really cool! I'm pretty amazed that storing cookies is so easy!
> > > > > > What happens if the local salt and the on-chain decrypted salt disagree? Which one is shown? Which one is used during reveal? Is it possible for the displayed salt to be correct, but the one used during reveal to be the incorrect, decrypted salt?
> > > > > > The reason I ask this is that I feel it might be confusing for us to _show_ the backup salt that the dApp _isn't_ actually using. It may be useful to get some UX guidance from @hlambur or @allisonlu22 here because I'm not entirely sure how the backup flow should work.
> > > > > > One alternative might be to store a history of salts/prices and show those backups in a modal if the user clicks a button. We can suggest that they click that button only in the case where they're having issues.
> > > > > > Thoughts?
> > > > > 
> > > > > 
> > > > > I only store the salt(s) if the `batchCommitFunction` reads "success", i.e. Drizzle sees that the transaction has gone through successfully. So the on-chain salt and the local salt shouldn't differ...right?
> > > > 
> > > > 
> > > > Isn't the idea of the backup that it protects you in case they do differ for some reason or the decryption fails? Or maybe I'm misunderstanding?
> > > 
> > > 
> > > I agree with you that the backup of the salt protects you in case the decryption (during the reveal phase) fails. This is what happened to me: the dApp did not decrypt the correct salt for my commit and therefore I couldn't reveal. What I would have liked to have done would be to use the salt to manually reveal through etherscan or a script.
> > 
> > 
> > I totally agree that having a backup salt would've been helpful in your case. I think we should be storing them as you're currently doing, but I think the presentation to the user is currently a little confusing. I think it's confusing because the backups are mixed in with other information (like the decrypted price) that the dApp _is_ actively using. This means that on the same line we're showing the _decrypted_ price and the _backup_ salt, leaving out the _decrypted_ salt and the (nonexistent) _backup_ price.
> > I was thinking an alternative might be a button on each line of the table that you click to show your local salt/price backups in a modal (we could even show all the backups you've ever made for that request. Then, we could make it obvious that if you want to use these, then you'll need to do the reveal manually.
> > Another, simpler option would be to just re-label the column as local salt backup, or something like that and explain in the tooltip that the dApp isn't using this salt.
> 
> I see what you're saying. I now save a local copy of the committed price and salt which should be available even if the app cannot decrypt the vote on-chain. I opted to still just save the latest committed data, which will only update if the `batchCommitFunction` succeeds.
> 
> Screen Shot 2020-05-28 at 15 24 17
> 
> @chrismaree unfortunately I had to scrap the copy button, as nice as it was UI-wise, because for some reason it doesn't work the way I intend it to

This is great, @nicholaspai. I think this UI is a lot clearer.

File: voter-dapp/src/ActiveRequests.js [hunk 3]
```diff
@@ -263,15 +288,46 @@ function ActiveRequests({ votingAccount, votingGateway }) {
...
   };
 
+  // Once transaction successfully sends, store latest batch of prices and salts in a cookie.
+  if (commitStatus === "success" && Object.keys(commitBackup).length > 0) {
+    for (let key in commitBackup) {
+      setCookie(key, commitBackup[key], { path: "/" });
+      // Delete object mapping so that we don't store it again.
+      delete commitBackup[key];
+    }
+  }
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js:329]: Yeah this is totally an edge case good point. If the user approves the 1st transaction, nothing ends up getting stored in the cookie

[the remark this answers is not in the document]
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js:329]: Drizzle also does not look handle this properly:

If user submits a transaction that is not the most recent one, then the `commitStatus` does not get set properly to "success" after a call

[the remark this answers is not in the document]
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js:329]: I might just simplify the logic and save every committed data, whether the user submits or not

[the remark this answers is not in the document]
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js]: As discussed with Matt IRL, I decided to keep a backup for all encrypted data, whether it is submitted on-chain or not.

File: voter-dapp/src/ActiveRequests.js [hunk 4]
```diff
@@ -374,6 +431,44 @@ function ActiveRequests({ votingAccount, votingGateway }) {
     }
   };
 
+  const copyStringToClipboard = string => {
```
Comment (mrice32) [inline voter-dapp/src/ActiveRequests.js:434]: Didn't you say that the clipboard copy didn't work?

File: voter-dapp/src/ActiveRequests.js [hunk 5]
```diff
@@ -263,10 +281,38 @@ function ActiveRequests({ votingAccount, votingGateway }) {
...
+      const newCommitKey = toVotingAccountAndPriceRequestKey(
+        votingAccount,
+        pendingRequests[index].identifier,
+        pendingRequests[index].time
+      );
+      const updatedCommitBackups = Object.assign(
+        {},
+        {
+          ...cookies[newCommitKey],
+          [Date.now()]: {
```
Comment (mrice32) [inline voter-dapp/src/ActiveRequests.js:300]: nit: maybe compute `Date.now()` outside the loop, so all the commits for a particular button-click have the same timestamp?

File: voter-dapp/src/ActiveRequests.js (hunk 4 above)
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js:434]: I got it to work by using the `navigator.writeText` object. Most browsers use the synchronous `document.execCommand` pattern, but Chrome replaced it with the asynchronous `navigator` command after version 66. There's a lengthy discussion [here](https://stackoverflow.com/questions/400212/how-do-i-copy-to-the-clipboard-in-javascript) it's pretty fascinating

File: voter-dapp/src/ActiveRequests.js (hunk 5 above)
Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js]: done

**Comment by nicholaspai** [comment]
@mrice32 @chrismaree I also added a tool tip to the "Current Vote" column to specify to the user that this is their most recent vote that we decrypted from their on-chain data. It is supposed to be, but possibly not, their local input price

Screen Shot 2020-05-29 at 10 31 16

File: voter-dapp/src/ActiveRequests.js (hunk 4 above)
Comment (mrice32) [reply voter-dapp/src/ActiveRequests.js:438]: Oh nice!

File: voter-dapp/src/ActiveRequests.js [hunk 6]
```diff
@@ -395,7 +520,28 @@ function ActiveRequests({ votingAccount, votingGateway }) {
             </TableCell>
             <TableCell className={classes.tableHeaderCell}>Timestamp</TableCell>
             <TableCell className={classes.tableHeaderCell}>Status</TableCell>
-            <TableCell className={classes.tableHeaderCell}>Current Vote</TableCell>
+            <TableCell className={classes.tableHeaderCell}>
+              Current Vote
+              <Tooltip
+                title="This is your most recently committed price for this request. We decrypt this value from your on-chain encrypted commit and your private key."
```
Comment (mrice32) [inline voter-dapp/src/ActiveRequests.js:526]:
```suggestion
                title="This is your most recently committed price for this request. We decrypt this value from your on-chain encrypted commit using your private key signature."
```

File: voter-dapp/src/ActiveRequests.js [hunk 7]
```diff
@@ -374,6 +435,44 @@ function ActiveRequests({ votingAccount, votingGateway }) {
...
+    }
+
+    // Asynchronously copy. This has been the way to copy to the clipboard for Chrome since v66.
+    navigator.clipboard.writeText(string).then(
+      function() {
+        console.log(`Async: Copied to clipboard`);
+      },
+      function(err) {
+        console.error("Async: Could not copy text: ", err);
+      }
```
Comment (mrice32) [inline voter-dapp/src/ActiveRequests.js:472]: Should this be in an `else`? I assumed that this code should only be run if we don't use the other method for copying, but maybe I'm misunderstanding?

Comment (nicholaspai) [reply voter-dapp/src/ActiveRequests.js:472]: yeah you're right. Either synchronously copy or async copy

**Pull request merged**